### PR TITLE
add new rfc 0032-bazel proposing bazelization of coda

### PR DESCRIPTION
**Pull request by mobileink**
Signed-off-by: Gregg Reynolds <601396+mobileink@users.noreply.github.com>

Thank you for contributing to Coda! Please see `CONTRIBUTING.md` if you haven't
yet.

Submitting draft RFC per instructions from bkase

### CODE REVIEW

File: rfcs/0031-bazel.md [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+target platform, toolchain, etc.  In particular, it should be easy to
+introduce new build tools.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+Platform support:
+  * Linux - required, I presume
+  * MacOS - my development environment
+  * Windows - this is a policy decision.  I can design for
```
Comment (bkase) [inline rfcs/0031-bazel.md:129]: Windows builds should be out-of-scope for now. This can be a separate RFC and a project for later.

MacOS and Linux are both required :+1:

File: rfcs/0031-bazel.md [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+* C/C++
+  * External libs. Most of these are already bazelized (minimal build on MacOS).
+    * jemalloc, libffi, libgmp, libomp, libpatch, libpq, libprocps, libre2, libsodium, openssl, zlib
+  * libsnark.
+    * A special case.  Coda uses a modified fork of libsnark; it
+      depends on a handful of C/C++ libs (libff, libfqfft, etc.).
+    * I have forked the dependencies on github and added Bazel support
+      to each (about 90% complete).  The ideal is to get the
+      maintainers to accept the Bazel-enabling patches.
+    * For Coda this approach may or may not be desirable.  The
```
Comment (bkase) [inline rfcs/0031-bazel.md:176]: While we wait for patches to be upstreamed on the internal deps and external BUILD file seems like a good option.

File: rfcs/0031-bazel.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+Alternatives
+* Stick with current situation.  That might work for o(1) internally
+  but is not so attractive for outside volunteers who would have to
+  struggle with the build system.  My guess is that improving the
+  current system would probably be more expensive in the long run than
+  switching to Bazel.
+* Nix may be an alternative, but according to the folks who provide
+  Bazel support for Haskell, Bazel and Nix work well together: [NIX +
+  BAZEL = FULLY REPRODUCIBLE, INCREMENTAL
+  BUILDS](https://www.tweag.io/posts/2018-03-15-bazel-nix.html).
```
Comment (bkase) [inline rfcs/0031-bazel.md:312]: Yes, the Nix/Bazel hybrid would work really well. ie. Bazel for the core project build, Nix for the deps on the outsides. We did this at Pinterest for our iOS builds and the combo worked very well. Of course, keep this part out of scope for now.

File: rfcs/0031-bazel.md [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+([OpenOCF](https://github.com/OpenOCF/iochibity)) from SCons to Bazel.
+Though the project was never completed, the switch to Bazel was, and
+was (in my opinion) quite successful.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What parts of the design do you expect to resolve through the RFC process before this gets merged?
+
+  * Which platforms and toolchains are to be officially supported.
```
Comment (bkase) [inline rfcs/0031-bazel.md:330]: In the short-term:

Linux and MacOS.

We should target x86_64 (and taking advantage of whichever SSE/AVX extensions we have turned on in the current CMake invocation for snarky). 

It would be preferable to support a Bazel toolchain via Nix for reproducibility's sake (as explained in https://www.tweag.io/posts/2018-03-15-bazel-nix.html ). But we should do this as a follow-up.

Initially, for macOS, supporting the "default" Xcode toolchain + OCaml installed in whichever way is easiest would work. For Linux, a Debian based Docker container installing the toolchain will suffice (similar to our existing `Dockerfile-toolchain` pulling from `ocaml/opam2:debian-9-ocaml-4.08`).

File: rfcs/0031-bazel.md [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+    package that would do this.
+  * Distributed builds ([Remote Execution](https://docs.bazel.build/versions/master/remote-execution.html))
+  * [Remote caching](https://docs.bazel.build/versions/master/remote-caching.html)
+  * Implementation of feature-testing in Starlark.
+
+#### Known Unknowns
+
+- Will Bazel and the OCaml ecosystem play nice?  Bazel is design as a
+  polyglot system, and already supports a variety of languages, so we
+  have reason to be optimistic about this.
```
Comment (bkase) [inline rfcs/0031-bazel.md:347]: I also think this will be fine.

File: rfcs/0031-bazel.md [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+Note: I don't have a GPU, and I'm not sure if I can compile CUDA code
+without one.  If I can, task 15 will be completed by end of month.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+The learning curve for Bazel is little steep in places, plus it has
+its own languge, Starlark - yet another language to learn.  It's a
+Python dialect, but still, somebody on staff will probably have to be
+tasked with learning the stuff.  On the other hand, once the build
```
Comment (bkase) [inline rfcs/0031-bazel.md:278]: I wrote a lot of Skylark a few years ago + I am on the core team 👍 . No worries.

File: rfcs/0031-bazel.md [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+system is in place it will probably not require much further work,
+just like Makefiles are usually left alone once they stabilize.
+
+Another issue with Bazel is feature detection.  Bazel just builds
+stuff; it does not provide the feature-detection functionality
+supported by GNU autoconf tools.  So it has no analog to
+`./configure`; you have to explicitly tell it about all features.
+That means that a Bazel build, in cases where feature detection is
+required, must be complimented by a traditional `configure` file or
```
Comment (bkase) [inline rfcs/0031-bazel.md:287]: Right now, we don't really have good feature detection support in our build anyway, so I'm not too worried about this one.

File: rfcs/0031-bazel.md [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+| 11 | all | Analyze existing Bazel/OCaml libs | 4 | ongoing | |
+| 12 | all | Develop rules_ocaml | 12 |Sun 5/31 |
+| 13 | all | Implement Bazel support for Coda OCaml code | 16 | Sun 5/31 |
+
+**Golang, CUDA**
+
+| | Platform | Task | Effort | Due | Notes |
+| -- | ----------- | ----------- | ------ | ---------- | ---------------|
+| 14 | all | go-libp2p | 4 | W 5/27  |
+| 15 | all | CUDA code in Coda | 1 | TBD | cuda-fixnum, cuda-prover, etc. |
```
Comment (bkase) [inline rfcs/0031-bazel.md:267]: The cuda builds are not currently integrated in our codebase, so you don't need to worry about it!

File: rfcs/0031-bazel.md [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+  the libraries with it.  It will require a little more work and
+  testing, but basically we're in good shape as far as C/C++ cross
+  compilation is concerned.
+* Haven't really looked into Bazel-Ocaml yet.
+
+### Effort Estimates and Timeline
+
+Effort = estimated remaining work effort in hours.
+
+C/C++
```
Comment (bkase) [inline rfcs/0031-bazel.md:233]: Cross builds to Linux are great if it's easy enough. ARM and Android cross definitely not required (plus you'll have a _really_ hard time getting it to work with a newer version of the OCaml compiler)

File: rfcs/0031-bazel.md [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+      * Implement `rules_ocaml` using this integrational strategy
+  * Use the new `rules_ocaml` to implement bazel support for Coda codebase:
+    * External libs
+    * Internal OCaml code
+    * C/C++ interop code
+* Golang
+  * Coda uses the Golang implementation of libp2p. Google maintains an
+    official set of Bazel rules for Go ([rules_go](https://github.com/bazelbuild/rules_go), so this should be
+    straightforward.
+
```
Comment (bkase) [inline rfcs/0031-bazel.md:212]: There is hidden work here that I want to make explicit. We will soon have a large number of crypto-related Rust dependencies in the build (example: https://github.com/o1-labs/marlin ). This will likely be almost as easy as Go to integrate. See https://github.com/google/cargo-raze and https://github.com/bazelbuild/rules_rust

File: rfcs/0031-bazel.md [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+  * libsnark.
+    * A special case.  Coda uses a modified fork of libsnark; it
+      depends on a handful of C/C++ libs (libff, libfqfft, etc.).
+    * I have forked the dependencies on github and added Bazel support
+      to each (about 90% complete).  The ideal is to get the
+      maintainers to accept the Bazel-enabling patches.
+    * For Coda this approach may or may not be desirable.  The
+      alternative is to use the same standard approach as used by the
+      other external libs, namely an external BUILD file.
+  * Internal C/C++ code.  The Code codebase contains some OCaml code
```
Comment (bkase) [inline rfcs/0031-bazel.md:179]: Do we ever actually generate C/C++ code from OCaml? Can you point out where this code lives?

File: rfcs/0031-bazel.md [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+| 5 | Mac to Android | cross-platform build | 8 | W 5/20 |Reusing and adapting code from previous project |
+| 6 | linux | configure dev. env | 2 | Fri 5/22 |assuming a docker container can be used as a linux dev environment on Mac |
+| 7 | linux | C/C++ min. bld | 2 | Fri 5/22 |the Mac stuff should work with few changes |
+| 8 | windows | configure dev. env | 4 | TBD | setup VM image, install windows, install tools, etc. |
+| 9 | windows | C/C++ min. bld | 8 | TBD |paths, compiler flags, etc. will take some work, depending on toolchain |
+
+Notes:
+
+1. External libs:
+  * Assumption: what works on the mac will require little or no effort to work on Linux and Windows
```
Comment (bkase) [inline rfcs/0031-bazel.md:250]: true for Linux, it will not work easily on Windows -- it's not worth looking into that for now.

File: rfcs/0031-bazel.md [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+  * libpq (Postgresql) won't build (possible bug in Bazel); workaround is to just install it manually
+
+**OCaml**
+
+| | Platform | Task | Effort | Due | Notes |
+| -- | ----------- | ----------- | ------ | ---------- | ---------------|
+| 10 | all | OCaml/Dune learning curve | 4 | ongoing |  I can usually fake it, since all languages are alike, but darnit, sometimes I have to sit down and actually learn something. |
+| 11 | all | Analyze existing Bazel/OCaml libs | 4 | ongoing | |
+| 12 | all | Develop rules_ocaml | 12 |Sun 5/31 |
+| 13 | all | Implement Bazel support for Coda OCaml code | 16 | Sun 5/31 |
```
Comment (bkase) [inline rfcs/0031-bazel.md:260]: I would suggest scoping this step to (at least) a week after rules_ocaml

**Comment by bkase** [comment]
Great proposal! This would be awesome to get working!

File: rfcs/0031-bazel.md [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+  * MacOS first, then Linux, then Windows.
+* Adaptive builds (local monobuilds with platform-specific adaptations).
+  * A Bazel `platform()` is defined for each officially supported platform
+    * Platform definition includes architecture, OS, toolchain.
+  * Build specifications are adapted to each supported platform
+    * For example, the compilers for different platforms take
+      different options.  Bazel does a good job of automatically
+      setting the compile command for each platform, but in some cases
+      additional options should or must be specified.
+  * Bazel does not have `./configure` functionality; it cannot detect
```
Comment (jin) [inline rfcs/0031-bazel.md:149]: You can use repository rules with `configure = true` as a proxy for this. Repository rules have unlimited access to analyze the system and generate BUILD files containing `platform` / `toolchain` declarations dynamically.

Check out bazel's own C++ autoconfigure repo rule: https://github.com/bazelbuild/bazel/blob/dcd19067212bfe2fb7eb2c404c2aa0cf3850752b/tools/cpp/cc_configure.bzl#L79-L86

File: rfcs/0031-bazel.md [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+  * [Remote caching](https://docs.bazel.build/versions/master/remote-caching.html)
+  * Implementation of feature-testing in Starlark.
+
+#### Known Unknowns
+
+- Will Bazel and the OCaml ecosystem play nice?  Bazel is design as a
+  polyglot system, and already supports a variety of languages, so we
+  have reason to be optimistic about this.
+- Will it really be cheaper/more efficient etc. to rely on one generic
+  build tool, instead of mashing up multiple specialized tools?  Bazel
```
Comment (jin) [inline rfcs/0031-bazel.md:349]: You'll most probably discover certain limitations by the underlying tools (like determinism, timestamps, undeclared outputs, dynamic dependencies) while operating in Bazel's stricter environment, and have to wrap / fork / patch those tools to do what you want.

File: rfcs/0031-bazel.md (hunk 14 above)
Comment (mobileink) [reply rfcs/0031-bazel.md:149]: Thanks.  I came across https://github.com/bazelbuild/bazel/issues/8786 a few days ago but have not yet looked into it in detail.

File: rfcs/0032-bazel.md [hunk 16]
```diff
@@ -0,0 +1,427 @@
...
+  line or via a config file (`.bazelrc`, `--bazelrc=file`).
+
+#### Integrated testing
+
+Bazel has builtin support for test targets.
+
+#### Cross-platform builds
+
+- Linux target from Mac host is in scope.  This is needed in order to
+  create docker images on the mac.
```
Comment (bkase) [inline rfcs/0032-bazel.md:104]: Why is this needed to create docker images on mac? For example, `make build` works from within the `codaprotocol/coda:toolchain-eb73e64d524c906384e9dbb23e9d0b8ae2be9e9e` Docker image from macOS.

File: rfcs/0032-bazel.md [hunk 17]
```diff
@@ -0,0 +1,427 @@
...
+## Detailed design
+[detailed-design]: #detailed-design
+
+Platform support:
+  * Linux - Required.
+  * MacOS - Required.
+  * Windows - Out of scope for this proposal.
+  * Docker - design of docker containers is out of scope, but support
+    for docker on Linux and Mac is in scope; this implies support
+    cross-platform builds on a Mac host targeting Linux.
```
Comment (bkase) [inline rfcs/0032-bazel.md:136]: The O(1) core team can help when you're ready for the container!

File: rfcs/0032-bazel.md (hunk 16 above)
Comment (mobileink) [reply rfcs/0032-bazel.md:104]: To be more precise: needed to create the linux binaries that go in the docker image.  I.e. in order to create the whole schmear on a mac.

Comment (mobileink) [reply rfcs/0032-bazel.md:104]: OTOH, presumably one could build the linux code in a linux docker image running on the mac, so I guess this should be downgraded from Required to Desirable but not Mandatory.
